api/v1alpha1/ndb_api_helpers.go

```diff
@@ -155,43 +156,119 @@ func GetNoneTimeMachineSLA(ctx context.Context, ndbclient *ndbclient.NDBClient)
-func GetOOBProfiles(ctx context.Context, ndbclient *ndbclient.NDBClient, dbType string) (profileMap map[string]ProfileResponse, err error) {
-	// Map of profile type to profiles
-	profileMap = make(map[string]ProfileResponse)
+// +kubebuilder:object:generate:=false
+type ProfileResolver interface {
+	Resolve(ctx context.Context, allProfiles []ProfileResponse, filter func(allProfiles ProfileResponse) bool) (profile ProfileResponse, err error)
+}
+
+func (providedProfile *Profile) Resolve(ctx context.Context, allProfiles []ProfileResponse, filter func(allProfiles ProfileResponse) bool) (profile ProfileResponse, err error) {
+	log := ctrllog.FromContext(ctx)
+	log.Info("Entered ndb_api_helpers.resolve", "provided profile", providedProfile)
+
+	// If both NAME and ID are not provided, return OOB profile
+	if providedProfile.Name == "" && providedProfile.Id == "" {
+		log.Info("Attempting to resolve the OOB profile...")
+		profile, err = util.FindFirst(allProfiles, filter)
+
+		if err != nil {
+			log.Error(err, "Error resolving OOB Profile, no profile information was provided in the spec")
+			return profile, errors.New("no OOB profile found")
+		}
+	} else { // Attempting to resolve using NAME
+		profile, _ = util.FindFirst(allProfiles, func(pr ProfileResponse) bool {
+			return pr.Name == providedProfile.Name
+		})
+
+		log.Info("Profile resolved using Name", "profile", profile)
 
-	profiles, err := GetAllProfiles(ctx, ndbclient)
+		// If unable to resolve by Name, then resolve using ID
+		if profile == (ProfileResponse{}) {
+			profile, err = util.FindFirst(allProfiles, func(pr ProfileResponse) bool {
+				return pr.Id == providedProfile.Id
+			})
 
+			log.Info("Profile found for the ID", "profile", profile)
+		}
+
+		// If neither worked, throw an error
+		if profile == (ProfileResponse{}) {
+			log.Error(err, "Error resolving profile by Name or Id")
+			return ProfileResponse{}, fmt.Errorf("could not resolve the profile by Name or Id, err=%d", err)
+		}
+	}
+
+	log.Info("Resolved the profile as", "profile", profile)
+	return profile, err
+}
+
+// TODO: Once the database_types refactoring is over, move out below methods to  profile_helper.go
+var ComputeOOBProfileResolver = func(p ProfileResponse) bool {
+	return p.Type == PROFILE_TYPE_COMPUTE && strings.Contains(strings.ToLower(p.Name), "small")
+}
+
+var SoftwareOOBProfileResolver = func(p ProfileResponse) bool {
+	return p.Type == PROFILE_TYPE_SOFTWARE && p.Topology == TOPOLOGY_SINGLE
+}
+
+var NetworkOOBProfileResolver = func(p ProfileResponse) bool {
+	return p.Type == PROFILE_TYPE_NETWORK
+}
+
+var DbParamOOBProfileResolver = func(p ProfileResponse) bool {
+	return p.Type == PROFILE_TYPE_DATABASE_PARAMETER
+}
+
+// Fetches all the profiles and returns a map of profiles
+// Returns an error if any profile is not found
+func GetProfiles(ctx context.Context, ndbclient *ndbclient.NDBClient, instanceSpec Instance) (profilesMap map[string]ProfileResponse, err error) {
```

Comment:
Would it make sense to call it ResolveProfiles?

Reply:
As discussed offline, it would be better to stay consistent and stick to the "Get" prefix. Resolving.